### PR TITLE
Implement native module states

### LINKED ISSUE #1586: Add different states for native module instead of using multiple bools
### Describe the bug

Currently native module uses multiple bools to track the state. We should ideally merge them to a single state and use state transitions.

Reference:
https://github.com/microsoft/ebpf-for-windows/pull/1547#discussion_r1015903298

### OS information

_No response_

### Steps taken to reproduce bug

NA

### Expected behavior

NA

### Actual outcome

NA

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by gtrevi**
## Description
Closes #1586.

This PR addresses unifying eBPF native module's multiple bool state flags, in order to provide more clarity.

### CODE REVIEW

**Comment by saxena-anurag** [comment]
> > Although the various eBPF native module's bool flags can't exactly be represented by a SM (as they are used simultaneously, i.e. detaching/unloading), this PR proposes to group them as bit fields, in order to provide more clarity.
> 
> If detaching and unloading are separate, that's fine but most of these are not. We should be able to combine at least 2 or 3 of them into one state enum.
> 
> ```
>         uint32_t initialized : 1;
>         uint32_t loading : 1;
>         uint32_t loaded : 1;
>         uint32_t detaching : 1;
>         uint32_t unloading : 1;
> ```
> 
> If Detaching and Unloading can happen in either order, then we could have multiple states like:
> 
> ```
> typedef enum {
>     Created = 0,
>     Initialized,
>     Loading,
>     Loaded,
>     Detaching,
>     Unloading,
>     DetachingAndUnloading
> } ebpf_native_module_state_t;
> ```

I think we can merge the following 3 in one state:
1. initialized -- set to true when module is loaded as part of LOAD_NATIVE_MODULE IOCTL
2. loading -- set to true when LOAD_NATIVE_PROGRAMS IOCTL is in progress.
3. loaded -- set to true when LOAD_NATIVE_PROGRAMS IOCTLS is completed.

And keep unloading and detaching as flags (use a single flag variable for both of these).

**Comment by gtrevi** [comment]
> > > Although the various eBPF native module's bool flags can't exactly be represented by a SM (as they are used simultaneously, i.e. detaching/unloading), this PR proposes to group them as bit fields, in order to provide more clarity.
> > 
> > 
> > If detaching and unloading are separate, that's fine but most of these are not. We should be able to combine at least 2 or 3 of them into one state enum.
> > ```
> >         uint32_t initialized : 1;
> >         uint32_t loading : 1;
> >         uint32_t loaded : 1;
> >         uint32_t detaching : 1;
> >         uint32_t unloading : 1;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If Detaching and Unloading can happen in either order, then we could have multiple states like:
> > ```
> > typedef enum {
> >     Created = 0,
> >     Initialized,
> >     Loading,
> >     Loaded,
> >     Detaching,
> >     Unloading,
> >     DetachingAndUnloading
> > } ebpf_native_module_state_t;
> > ```
> 
> I think we can merge the following 3 in one state:
> 
> 1. initialized -- set to true when module is loaded as part of LOAD_NATIVE_MODULE IOCTL
> 2. loading -- set to true when LOAD_NATIVE_PROGRAMS IOCTL is in progress.
> 3. loaded -- set to true when LOAD_NATIVE_PROGRAMS IOCTLS is completed.
> 
> And keep unloading and detaching as flags (use a single flag variable for both of these).

@dthaler the full enum was in the first commit (which failed unit tests), and there is not just one combination of bools being used simultaneously.

@saxena-anurag tests also failed limiting the enum to just the three [uninitialized, initialized, loading, loaded]. I was able to isolate just [uninitialized, initialized, loaded] (which is not much value).

In thinking that mix-matching enum & bool statuses is not ideal, the full the bit-field looked like the most clear and synthetic representation.
Happy to explore further options though.

**Comment by dthaler** [comment]
> @dthaler the full enum was in the first commit (which failed unit tests), and there is not just one combination of bools being used simultaneously.

This one is not a big deal so I wouldn't want to spend too much time on it, but I don't understand why an enum like I mentioned would fail tests.  What other bools are used simultaneously other than detaching and unloading?   How difficult would it be to fix the design to not have two bools used simultaneously but to have a fixed state machine (e.g., force detaching to come before unloading)?

**Comment by gtrevi** [comment]
> > @dthaler the full enum was in the first commit (which failed unit tests), and there is not just one combination of bools being used simultaneously.
> 
> This one is not a big deal so I wouldn't want to spend too much time on it, but I don't understand why an enum like I mentioned would fail tests. What other bools are used simultaneously other than detaching and unloading? How difficult would it be to fix the design to not have two bools used simultaneously but to have a fixed state machine (e.g., force detaching to come before unloading)?

A slight tweak to the initialization check should have fixed, leaving just "detaching" as a standalone flag. This also highlighted that we should have a new negative-test case for loading native programs without module initialization -> opened #1660 to track.

File: libs/execution_context/ebpf_native.c [hunk 1]
```diff
@@ -42,18 +42,23 @@ typedef struct _ebpf_native_module_cleanup
     ebpf_preemptible_work_item_t* cleanup_workitem;
 } ebpf_native_module_cleanup_t;
 
+typedef enum _ebpf_native_module_state
+{
+    MODULE_STATE_UNINITIALIZED = 0,
+    MODULE_STATE_INITIALIZED,
+    MODULE_STATE_LOADING,
+    MODULE_STATE_LOADED,
+    MODULE_STATE_UNLOADING,
```
Comment (saxena-anurag) [inline libs/execution_context/ebpf_native.c:51]: We should use lower_snake_case for the enums.

https://github.com/microsoft/ebpf-for-windows/blob/b599ee24fd32582c659d33cbc46310b828fd5315/docs/DevelopmentGuide.md#naming-conventions

File: libs/execution_context/ebpf_native.c [hunk 2]
```diff
@@ -1148,18 +1154,18 @@ ebpf_native_load(
             *module_id);
         goto Done;
     }
-    if (module->detaching || module->unloading) {
-        // This client is already detaching / unloading.
+    if (module->detaching || module->state == MODULE_STATE_UNLOADING) {
```
Comment (saxena-anurag) [inline libs/execution_context/ebpf_native.c:1157]: I think we should merge the if check at line 1146 and line 1157 and set the result based on the current module state.

File: libs/execution_context/ebpf_native.c [hunk 3]
```diff
@@ -1222,7 +1228,7 @@ ebpf_native_load_programs(
     module = *existing_module;
     module_state = ebpf_lock_lock(&module->lock);
     native_lock_acquired = true;
-    if (module->loading || module->loaded) {
+    if (module->state == MODULE_STATE_LOADING || module->state == MODULE_STATE_LOADED) {
```
Comment (saxena-anurag) [inline libs/execution_context/ebpf_native.c:1231]: This should be `if (module->state != MODULE_STATE_INITIALIZED || module->detaching)`

File: libs/execution_context/ebpf_native.c [hunk 4]
```diff
@@ -1233,7 +1239,7 @@ ebpf_native_load_programs(
         goto Done;
     }
 
-    if (module->unloading || module->detaching) {
+    if (module->detaching || module->state == MODULE_STATE_UNLOADING) {
```
Comment (saxena-anurag) [inline libs/execution_context/ebpf_native.c:1242]: As earlier, I think we can merge this if check and the one at line 1231.

**Comment by saxena-anurag** [comment]
This PR contains ubpf submodule changes also. I think they should be a separate PR.

**Comment by gtrevi** [comment]
> This PR contains ubpf submodule changes also. I think they should be a separate PR.

That must be a pointer issue, not a change, will revert that

File: libs/execution_context/ebpf_native.c (hunk 1 above)
Comment (gtrevi) [reply libs/execution_context/ebpf_native.c:51]: Heads-up that at that link, it says (I didn't see any exception for enum types):

```
2. Use UPPER_SNAKE_CASE for macro names and constants.
```

Therefore, in mimicking how enum types are defined throughout the whole codebase, I implemented that way:

e.g. `ebpf_base.h`
```c
typedef enum _ebpf_return_type {
    EBPF_RETURN_TYPE_INTEGER = 0,
    EBPF_RETURN_TYPE_PTR_TO_MAP_VALUE_OR_NULL,
    EBPF_RETURN_TYPE_INTEGER_OR_NO_RETURN_IF_SUCCEED,
    EBPF_RETURN_TYPE_UNSUPPORTED,
} ebpf_return_type_t;
typedef enum _ebpf_argument_type {
...
}
```

Happy to change it to whatever best for this specific case though, please let me know.

File: libs/execution_context/ebpf_native.c (hunk 4 above)
Comment (gtrevi) [reply libs/execution_context/ebpf_native.c:1242]: @saxena-anurag Sure fine with that, just want to ensure we really want to merge the current distinction between `EBPF_OBJECT_ALREADY_EXISTS` and `EBPF_EXTENSION_FAILED_TO_LOAD` (should any existing code rely on this).
If so, what unified error would be more appropriate?

File: libs/execution_context/ebpf_native.c (hunk 2 above)
Comment (gtrevi) [reply libs/execution_context/ebpf_native.c:1157]: (comment to same suggestion below)

Comment (saxena-anurag) [reply libs/execution_context/ebpf_native.c:1157]: We can still set different return values based on the current module state. But will be better to merge the state check under one condition.

Also, per current PR changes, if the state is `MODULE_STATE_UNLOADING`, this function will bail from line 1155 and never hit line 1157.

File: libs/execution_context/ebpf_native.c (hunk 1 above)
Comment (dthaler) [reply libs/execution_context/ebpf_native.c:51]: UPPER_SNAKE_CASE is correct for enum values

Comment (gtrevi) [reply libs/execution_context/ebpf_native.c:51]: It's fine for me, as that's exactly how I interpreted it.

**Review by saxena-anurag — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

**Pull request merged**